File: packages/core/src/contracts/resource.ts
import type { CapabilityMetadata } from "./channel.js";

export interface IResource<TInput = unknown, TOutput = unknown> {
  readonly metadata: CapabilityMetadata;
  execute(input: TInput, context: ResourceContext): Promise<TOutput>;
}

export interface ResourceContext {
  pipelineId?: string;
  workdir?: string;
}

File: crates/poria-core/src/contracts/resource.rs
use async_trait::async_trait;

use super::channel::CapabilityMetadata;

#[derive(Debug, Clone)]
pub struct ResourceContext {
    pub pipeline_id: Option<String>,
    pub workdir: Option<String>,
}

#[async_trait]
pub trait Resource: Send + Sync {
    fn metadata(&self) -> &CapabilityMetadata;
    async fn execute(
        &self,
        input: serde_json::Value,
        ctx: ResourceContext,
    ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>>;
}
